### PR TITLE
Halo optimization and solving the race condition on corner nodes

### LINKED ISSUE #1179: Increase ≈ tolerance to account for differences across GPU microarchitectures
After the big changes yesterday I decided to run the tests to make sure everything was working.  Thanks again @ali-ramadhan for helping me get that started.

Now I'm finding that there are some failures and somethings that are broken.  See below.  Is it just me or do others get this now?

```
Test Summary:                            | Pass  Fail  Broken  Total
Oceananigans                             | 2987     8       5   3000
  Unit tests                             | 1511             1   1512
  Model and time stepping tests (part 1) |   99                   99
  Model and time stepping tests (part 2) |  214             1    215
  Simulation tests                       | 1142     2       3   1147
    Simulations                          |   26                   26
    Diagnostics                          |   12                   12
    Output writers                       |  409     2            411
      FieldSlicer                        |    1                    1
      WindowedTimeAverage                |    2                    2
      NetCDF [GPU]                       |  198                  198
      JLD2 [GPU]                         |   11                   11
      Checkpointer [GPU]                 |  166     2            168
      Dependency adding [GPU]            |    2                    2
      Time averaging of output [GPU]     |   29                   29
    Abstract operations                  |  695             3    698
  Regression                             |   14     6             20
    Thermal bubble [GPU]                 |    5                    5
    Rayleigh–Bénard tracer [GPU]         |    5                    5
    Ocean large eddy simulation [GPU]    |    4     6             10
  Scripts                                |    7                    7
ERROR: LoadError: Some tests did not pass: 2987 passed, 8 failed, 0 errored, 5 broken.
in expression starting at /home/fpoulin/software/Oceananigans.jl/test/runtests.jl:77
ERROR: Package Oceananigans errored during testing
```

### PR DESCRIPTION
**Pull request by simone-silvestri**
This PR is created to solve the boundary condition race condition that occurs when trying to fill in the halo corner nodes.

It is basically just an integration of PR #1985 that runs the halo filling sequentially to avoid unsynchronized execution and PR #1923 to fill left and right halos together which should increase performance.

IMPORTANT NOTE:

This PR will fail the ocean large eddy simulation regression tests when lines (50-53) in BoundaryConditions/fill_halo_regions.jl are uncommented.

The only thing that these lines do is a reordering of halo filling order such that the periodic boundary conditions are calculated after all the "Bounded type" boundary conditions.

This is needed since periodic boundary conditions require previous evaluation of boundary conditions in other directions to fill the corner nodes (see PR #1985).

It is possible that the ocean-large-eddy-simulation-regression-data has been generated with periodic boundary conditions (west-east, north-south) which are evaluated before the bounded ones (top-bottom) (see the fill_halo_regions! function in main). As such the issue with the test failing when uncommenting lines (50-53) would not be a bug in the code but a bug in the previous generated regression data. This error is quite small (as the corners would be updated based on the previous step "bounded type" boundary conditions) but still enough to make the test fail. (I leave this to you to confirm :))

Resolves #1179

### CODE REVIEW

File: src/BoundaryConditions/fill_halo_regions.jl [hunk 1]
```diff
@@ -22,35 +22,90 @@ function fill_halo_regions!(fields::Union{Tuple, NamedTuple}, arch, args...)
...
+
+    return NoneEvent()
+end
+
+# Fallbacks split into two calls
+function fill_west_and_east_halo!(c, west_bc, east_bc, args...; kwargs...)
+     west_event = fill_west_halo!(c, west_bc, args...; kwargs...)
+     east_event = fill_east_halo!(c, east_bc, args...; kwargs...)
+    multi_event = (west_event, east_event)
+    return multi_event
```
Comment (glwagner) [inline src/BoundaryConditions/fill_halo_regions.jl:76]: This might be an improvement:

```suggestion
    return MultiEvent((west_event, east_event))
```

and the same for those below.

File: src/BoundaryConditions/fill_halo_regions_flux.jl [hunk 2]
```diff
@@ -1,66 +1,127 @@
...
+    i, j = @index(Global, NTuple)
+
+    @unroll for k in 1:grid.Hz
+        _fill_flux_bottom_halo!(i, j, k, grid, c)
+        _fill_flux_top_halo!(i, j, k, grid, c)
+    end
+end
+
+fill_west_and_east_halo!(c, west_bc::FBC, east_bc::FBC, arch, dep, grid, args...; kwargs...) =
+    (NoneEvent(), launch!(arch, grid, :yz, fill_flux_west_and_east_halo!, c, grid; dependencies=dep, kwargs...))
```
Comment (glwagner) [inline src/BoundaryConditions/fill_halo_regions_flux.jl:58]: If the "individual halo filling functions" return a `MultiEvent`, then I think this function (and all similar functions) can return

```suggestion
    launch!(arch, grid, :yz, fill_flux_west_and_east_halo!, c, grid; dependencies=dep, kwargs...)
```

what do you think @simone-silvestri ?

**Comment by glwagner** [comment]
Very nice @simone-silvestri . I think we have to regenerate the ocean LES regression test data. Unfortunately this regression test doesn't use `DataDeps` yet so regenerating it will increase the size of the repo by a little bit. Another option is to refactor the regression tests to use DataDeps but this might be best saved for a future PR.

I think this fix may actually resolve some ancient issues that we didn't understand until now, for example:

https://github.com/CliMA/Oceananigans.jl/issues/1179

To summarize:

When running a model that has both `Periodic` and `Bounded` directions, _and_ when using a stencil that touches corner points (such as with `coriolis isa FPlane`, the order in which halos are filled matters. In particular, we must fill periodic directions _last_, because the corners then depend on a halo point which is just outside the boundary in the `Bounded` direction(s).

The ocean large eddy simulation regression test is one such case (horizontally-periodic, bounded in z). There is a no-flux surface velocity boundary condition that creates a dependency between the bounded and periodic halo fills.

The previous algorithm assumed that all halos could be filled simultaneously with no dependencies on one another. However, because the CPU is serial the halos were nevertheless filled in the ordering prescribed by `fill_halo_regions!`:

https://github.com/CliMA/Oceananigans.jl/blob/2d6ccfe94ce2c68857f70b2f8839020049932e00/src/BoundaryConditions/fill_halo_regions.jl#L32-L42

In other words, the vertical halos (the bounded direction) are filled last on the CPU.

On the GPU the ordering may change because on the GPU KernelAbstractions is no longer serial. Thus on occasion the periodic halos might have been filled _after_ the bounded halos, leading to a failure of the regression test.

This PR fixes that issue so that periodic directions are always filled last. So I think we should expect that regression tests fail on both the CPU and GPU.

**Comment by simone-silvestri** [comment]
@glwagner Yes it does also fail on CPU!

File: src/BoundaryConditions/fill_halo_regions.jl [hunk 3]
```diff
@@ -51,7 +51,7 @@ function fill_halo_regions!(c::OffsetArray, field_bcs, arch, grid, args...; kwar
     # fill_halos! = fill_halos![perm]
     # field_bcs_array_left  = field_bcs_array_left[perm]
     # field_bcs_array_right = field_bcs_array_right[perm]
-
+#
```
Comment (glwagner) [inline src/BoundaryConditions/fill_halo_regions.jl:54]: 👀

File: src/BoundaryConditions/fill_halo_regions_flux.jl (hunk 2 above)
Comment (simone-silvestri) [reply src/BoundaryConditions/fill_halo_regions_flux.jl:58]: You're right, I ll change that

**Comment by glwagner** [comment]
We should run some similar benchmarks as were run on #1923 to see if this change yields a speed up (or not).

File: test/runtests.jl [hunk 4]
```diff
@@ -79,104 +79,104 @@ archs = test_architectures()
...
+    #         include("test_cubed_sphere_halo_exchange.jl")
+    #         include("test_cubed_sphere_circulation.jl")
+    #     end
+    # end
+
+    # if group == :distributed || group == :all
+    #     MPI.Initialized() || MPI.Init()
+    #     include("test_distributed_models.jl")
+    #     include("test_distributed_poisson_solvers.jl")
+    # end
```
Comment (glwagner) [inline test/runtests.jl]: Good way to get tests to pass.

File: test/runtests.jl [hunk 5]
```diff
@@ -79,104 +79,104 @@ archs = test_architectures()
...
-    if group == :scripts || group == :all
-        @testset "Scripts" begin
-            include("test_validation.jl")
-        end
-    end
+    # if group == :scripts || group == :all
+    #     @testset "Scripts" begin
+    #         include("test_validation.jl")
+    #     end
+    # end
```
Comment (glwagner) [inline test/runtests.jl]: 💪

**Comment by simone-silvestri** [comment]
This PR is almost ready to merge. Shall we agree to uncomment the following

```
## REMEMBER! NEW DATA HAS TO BE GENERATED FOR THE OCEAN AND HYDROSTATIC REGRESSION WITH THE REORDERING OF BC


# perm = sortperm(field_bcs_array_left, lt=fill_first)

# fill_halos! = fill_halos![perm]
# field_bcs_array_left  = field_bcs_array_left[perm]
# field_bcs_array_right = field_bcs_array_right[perm]
```

and regenerate regression data on this PR before merging?

**Comment by glwagner** [comment]
Beautiful! Let's do it.

**Comment by glwagner** [comment]
Some benchmarks...

## On `main`

```julia
julia> include("benchmark_two_dimensional_models.jl")
[ Info: Precompiling Oceananigans [9e8cae18-63c1-5223-a75c-80ca9d6e9a09]
[ Info: Oceananigans will use 24 threads
[ Info: Benchmarking CPU() model with RegularRectilinearGrid{Float64, Periodic, Periodic, Bounded}(Nx=256, Ny=256, Nz=1)...
  176.874 ms (66322 allocations: 23.79 MiB)
[ Info: Benchmarking CPU() model with RegularRectilinearGrid{Float64, Periodic, Periodic, Bounded}(Nx=256, Ny=1, Nz=256)...
  500.601 ms (111291 allocations: 276.81 MiB)
[ Info: Benchmarking CPU() model with RegularRectilinearGrid{Float64, Periodic, Periodic, Bounded}(Nx=1, Ny=256, Nz=256)...
  320.139 ms (48245 allocations: 269.93 MiB)
[ Info: Benchmarking GPU() model with RegularRectilinearGrid{Float64, Periodic, Periodic, Bounded}(Nx=256, Ny=256, Nz=1)...
  14.721 ms (36840 allocations: 12.11 MiB)
[ Info: Benchmarking GPU() model with RegularRectilinearGrid{Float64, Periodic, Periodic, Bounded}(Nx=256, Ny=1, Nz=256)...
  23.795 ms (40688 allocations: 12.18 MiB)
[ Info: Benchmarking GPU() model with RegularRectilinearGrid{Float64, Periodic, Periodic, Bounded}(Nx=1, Ny=256, Nz=256)...
  17.219 ms (38055 allocations: 12.14 MiB)
```

## This PR
julia> include("benchmark_two_dimensional_models.jl")
[ Info: Benchmarking CPU() model with RegularRectilinearGrid{Float64, Periodic, Periodic, Bounded}(Nx=256, Ny=256, Nz=1)...
  104.521 ms (77858 allocations: 25.03 MiB)
[ Info: Benchmarking CPU() model with RegularRectilinearGrid{Float64, Periodic, Periodic, Bounded}(Nx=256, Ny=1, Nz=256)...
  235.623 ms (120254 allocations: 26.49 MiB)
[ Info: Benchmarking CPU() model with RegularRectilinearGrid{Float64, Periodic, Periodic, Bounded}(Nx=1, Ny=256, Nz=256)...
  137.914 ms (72488 allocations: 19.41 MiB)
[ Info: Benchmarking GPU() model with RegularRectilinearGrid{Float64, Periodic, Periodic, Bounded}(Nx=256, Ny=256, Nz=1)...
  14.675 ms (37239 allocations: 11.65 MiB)
[ Info: Benchmarking GPU() model with RegularRectilinearGrid{Float64, Periodic, Periodic, Bounded}(Nx=256, Ny=1, Nz=256)...
  22.468 ms (41731 allocations: 11.71 MiB)
[ Info: Benchmarking GPU() model with RegularRectilinearGrid{Float64, Periodic, Periodic, Bounded}(Nx=1, Ny=256, Nz=256)...
  14.661 ms (37308 allocations: 11.65 MiB)
```

This PR makes substantial improvements on the CPU at 256^2  (all models are faster, and 2D models in xz or yz are more than 2x faster). For GPU I think the benchmark might need to be bigger (or smaller?!) to see differences.

**Review by glwagner — APPROVED** [review]
